tests/unittests/storage/test_storage.py

```diff
@@ -0,0 +1,363 @@
+    ]
+}
+
+
+def _generate(obj, *args, value):
+    if obj is None:
+        return None
+
+    obj = copy.deepcopy(obj)
+    data = obj
+
+    for arg in args[:-1]:
+        data = data[arg]
+
+    data[args[-1]] = value
+    return obj
+
+
+def make_lost_trial():
+    """Make a lost trial"""
+    obj = copy.deepcopy(base_trial)
+    obj['status'] = 'reserved'
+    obj['heartbeat'] = datetime.datetime.utcnow() - datetime.timedelta(seconds=61 * 2)
+    obj['params'].append({
+        'name': '/index',
+        'type': 'categorical',
+        'value': 'lost_trial'
+    })
+    return obj
+
+
+def generate_trials():
+    """Generate Trials with different configurations"""
+    status = ['completed', 'broken', 'reserved', 'interrupted', 'suspended', 'new']
+    new_trials = [_generate(base_trial, 'status', value=s) for s in status]
+
+    # make each trial unique
+    for i, trial in enumerate(new_trials):
+        trial['params'].append({
+            'name': '/index',
+            'type': 'categorical',
+            'value': i
+        })
+
+    return new_trials
+
+
+def generate_experiments():
+    """Generate a set of experiments"""
+    users = ['a', 'b', 'c']
+    return [_generate(base_trial, 'metadata', 'user', value=u) for u in users]
+
+
+@pytest.mark.parametrize('storage', storage_backends)
+class TestStorage:
+    """Test all storage backend"""
+
+    def test_create_experiment(self, storage):
+        """Test create experiment"""
+        with OrionState(experiments=[], database=storage) as cfg:
+            storage = cfg.storage()
+
+            storage.create_experiment(base_experiment)
+
+            experiments = storage.fetch_experiments({})
+            assert len(experiments) == 1, 'Only one experiment in the database'
+
+            experiment = experiments[0]
+            assert base_experiment == experiment, 'Local experiment and DB should match'
+
+    def test_fetch_experiments(self, storage, name='default_name', user='default_user'):
+        """Test fetch expriments"""
+        with OrionState(experiments=[base_experiment], database=storage) as cfg:
+            storage = cfg.storage()
+
+            experiments = storage.fetch_experiments({'name': name, 'metadata.user': user})
+            assert len(experiments) == len(cfg.experiments)
+
+            experiment = experiments[0]
+            assert experiment['name'] == name, 'name should match query'
+            assert experiment['metadata']['user'] == user, 'user name should match query'
+
+    def test_register_trial(self, storage):
+        """Test register trial"""
+        with OrionState(experiments=[base_experiment], database=storage) as cfg:
+            storage = cfg.storage()
+            trial1 = storage.register_trial(Trial(**base_trial))
+            trial2 = storage.get_trial(trial1)
+
+            assert trial1.to_dict() == trial2.to_dict(), 'Trials should match after insert'
+
+    def test_register_lie(self, storage):
```

Comment:
Should add a test where trial lie cannot be registered because of DuplicateKeyError.